### ISSUE TITLE
AdGuard DNS dashboard fails to recognize a connected device

### ISSUE THREAD

**Issue by kriswilk** [AdguardTeam/AdGuardDNS#850]
### Platform

Windows

### Protocol

DNS-over-HTTPS

### Do you use AdGuard app?

Yes I am

### Your configuration

- Windows 11 24H2
- Chrome 130.0.6723.92 (Official Build) (64-bit)
- Chrome Secure DNS disabled
- AdGuard 7.19.0 with HTTPS filtering enabled
- AdGuard Personal DoH URL entered into AdGuard

### Traceroute to AdGuard DNS

```
Tracing route to dns.adguard.com [94.140.14.14]
over a maximum of 30 hops:

  1     3 ms    <1 ms    <1 ms  10.0.0.1
  2    12 ms    13 ms    12 ms  pool-99-246-142-1.cpe.net.cable.rogers.com [99.246.142.1]
  3    14 ms    16 ms    12 ms  24.156.150.233
  4    12 ms    13 ms     5 ms  3032-cgw01.wlfdle.rmgt.net.rogers.com [209.148.232.37]
  5    13 ms    12 ms    53 ms  209.148.235.214
  6     *        *        *     Request timed out.
  7    14 ms    13 ms    13 ms  38.32.56.42
  8    17 ms    11 ms    12 ms  vl204.tor-clg1-dist-2.cdn77.com [185.229.188.243]
  9    11 ms    12 ms    14 ms  dns.adguard.com [94.140.14.14]

Trace complete.
```

### Issue Details

**TL;DR:** AdGuard DNS doesn't seem to recognize when it's connected to at least some of its own DNS servers.

On multiple systems (including my mobile device running AdGuard on Android 9), the AdGuard DNS fails to recognize that I've configured AdGuard DNS within AdGuard.

The steps I followed on the system config listed above were:

1. Confirm that AdGuard is active and set to use my AdGuard Personal DoH URL.
2. Visit some web sites to generate some DNS requests.
3. Visit the AdGuard DNS dashboard.
4. Inspect the query log for the device to confirm that its requests are being resolved by AdGuard DNS.
5. Observe the top banner reporting the connection status.
6. Visit the AdGuard test page (https://adguard.com/en/test.html).

The device is clearly connected to AdGuard DNS and seems to be working correctly but AdGuard claims that it's not. Secure DNS is disabled in Chrome, so this is not the issue.

Furthermore, substituting my NextDNS DoH URL works perfectly fine on the NextDNS portal. It knows I'm connected immediately.

### Expected Behavior

1. The dashboard should report new requests/blocking for the connected device.
2. The dashboard should indicate that the device is "CONNECTED" to AdGuard DNS.
3. The test page should indicate that AdGuard DNS is "RUNNING".

### Actual Behavior

1. The dashboard CORRECTLY shows all DNS traffic, proving that the device is sending DNS requests to AdGuard DNS.
2. The dashboard ERRONEOUSLY states "This device is not added to AdGuard DNS. Connect it to enable DNS protection."
3. The dashboard ERRONEOUSLY states that AdGuard DNS is "Not running"

### Screenshots

![Screenshot 2024-11-05 112658](https://github.com/user-attachments/assets/02e6a4b7-b0ad-43f8-9eda-3136fd27a8fc)
![Screenshot 2024-11-05 112750](https://github.com/user-attachments/assets/20cc4dc0-df88-44a3-aaa9-88a4f49f4c42)
![Screenshot 2024-11-05 112856](https://github.com/user-attachments/assets/f31e6b94-64a1-4858-bbb0-e02010c48460)

**Comment by kriswilk** [comment]
Another very strange issue, maybe related: despite the fact that I am near Toronto, Canada, the DNS servers selected seem to be in Brazil. Shouldn't I be connecting to the Toronto server(s)??

**Comment by kriswilk** [comment]
Upon further testing, I also discovered that:

- using the QUIC address, the AdGuard dashboard and test pages CORRECTLY detect that I'm connected to AdGuard DNS.
- using the HTTPS or TLS URLs, it is reported that I'm not connected (even though I am...).

In all cases, the servers handling requests seem to be in Brazil rather than local options. And again, ALL of these URLs work fine when I enable Chrome Secure DNS and paste them directly into the browser. The issue is ONLY when I configure DNS in the AdGuard app.

**Comment by ameshkov** [comment]
This all is very strange.

I'd understand if it was only happening on your Win device (maybe there's some other software intercepting DNS after AG for instance), but I cannot explain the fact that it's also happening on Android.

Anyways, could you please record debug-level log in AdGuard and also send it to us? We'll try to figure it that way.

https://adguard.com/kb/adguard-for-windows/solving-problems/adguard-logs/#debug-logs

**Comment by kriswilk** [comment]
I just sent a debug log in the same email ticket. Around 2:35 pm (Toronto/Eastern Time) I did this:

- Confirmed DoH URL in AdGuard settings
- Confirmed Chrome Secure DNS is DISABLED
- Tried loading the AdGuard test page twice. Both times reported AdGuard DNS "not running".
- Tried loading the AdGuard DNS dashboard. Status indicated "device not added to AdGuard DNS".
- Changed AdGuard settings to use DoQ URL. Waited a few seconds.
- Repeated the same page load tests. All cases now show "running" or "device XXXX is protected..."

I hope something in this helps. I am confident that my requests are being filtered as all dashboard analytics show my activity. But these connection tests are having some kind of trouble.

**Comment by emeritaacuity0u** [comment]
@kriswilk - As a curiosity, what does [dnscheck.tools](https://www.dnscheck.tools) show your DNS resolvers to be?

**Comment by kriswilk** [comment]
@emeritaacuity0u, here is a very concise summary in two screenshots. Nothing was changed except the DNS URL type. Curiously, on HTTPS I now get servers in Toronto, and on QUIC I get a bunch of servers in Brazil. Strange routing choices, but this does demonstrate that the problem is not with the Brazil servers (I thought it might be). Anyway, the choice of routing seems really strange.

USING DoH (**_not_** recognized by AdGuard DNS dashboard):
![Screenshot 2024-11-05 190438](https://github.com/user-attachments/assets/bc5d8369-0188-48e4-8688-c6323bbcc9c4)

USING DoQ (recognized by AdGuard DNS dashboard):
![Screenshot 2024-11-05 190226](https://github.com/user-attachments/assets/d5b43d9b-4250-4e00-b889-22e2fa9c4a5d)

**Comment by ameshkov** [comment]
Argh, finally I figured it out! It does indeed have something to do with the routing.

Just so that you understood the idea:

* `{random}-dnscheck.adguard-dns.com` domain can only be resolved via AdGuard DNS and the fact that you're getting a 404 in response means that it was actually resolved and you used AdGuard DNS server for that.
* When this domain is being resolved we "store" the `{random}` for a short period of time so that when the HTTP request from the browser arrives we could pull it and respond with the information recorded when we the DNS query was processed.
* `404` can be returned when the `{random}` string from the hostname was not found, but if we received a DNS query it must've been recorded.

The problem is that in the current setup the storage for `{random}` values is one per data center, i.e. if we receive a DNS query in Toronto, a server in Brazil won't be able to pull the data from there.

And this is exactly what happens:
1. AdGuard for Win DoH client prefers IPv4 so your queries go to Toronto.
2. But then Chrome prefers IPv6 and the `dnscheck` HTTP request goes to Brazil.

Changing routing on our own is close to impossible, especially with IPv6, so here we can only hope that at some point it will improve.

The good news is that we're about to roll out a new global KV storage and replace the old one limited to a single data-center. 

Maybe we'll even be able to do it today so stay tuned.

**Comment by kriswilk** [comment]
> Argh, finally I figured it out! It does indeed have something to do with the routing.
> 
> . . .
> 
> The good news is that we're about to roll out a new global KV storage and replace the old one limited to a single data-center.

I'm glad to hear that I am not going crazy! I look forward to the update. If you comment here when the change(s) are made I will test again on all my devices. 👍

**Comment by ameshkov** [comment]
@kriswilk could you please check it again now? Should be fixed already

**Comment by kriswilk** [comment]
> @kriswilk could you please check it again now? Should be fixed already

I just did this:

- Set AdGuard for Windows to use AdGuard public DNS for simplicity
- Flushed both browser and system DNS caches
- Rebooted my PC
- Opened DNS Check and AdGuard Test Page in a fresh Edge profile.
- Simultaneously opened the same pages in Windows Sandbox also in Edge.

Single screenshot shows everything. Left side is my native Windows, right side is the Windows Sandbox running at the same time. Main difference is the sandbox doesn't have AdGuard running natively.

![Screenshot 2024-11-06 100609](https://github.com/user-attachments/assets/7449e3d3-677e-4425-867e-f16a05c13768)
![Screenshot 2024-11-06 100553](https://github.com/user-attachments/assets/bc413c25-68ab-4dac-8df5-63c449db2047)

**Comment by ameshkov** [comment]
Oh no:(

Let's at least try verifying my initial idea just in case: please try disabling IPv6 in Windows network settings and see if it changes anything.

**Comment by kriswilk** [comment]
Everything was still on-screen when you replied, so I disabled IPv6 on all network adapters and refreshed all pages. Indeed the problem is resolved that way:

![Screenshot 2024-11-06 102604](https://github.com/user-attachments/assets/a14e76bb-0394-4165-bc21-3c65343f2a08)

**Comment by kriswilk** [comment]
It seems I am not the only one experiencing this:

https://www.reddit.com/r/Adguard/comments/1gjk9pf/comment/lvpuepb/

I linked to this issue so anyone else can come here instead of raising separate issues/tickets.

**Comment by ameshkov** [comment]
@kriswilk now it's fixed for good :)

**Comment by kriswilk** [comment]
> @kriswilk now it's fixed for good :)

YES, it seems to be! It's working correctly on several PCs and also on my phone.

Thanks for such quick responses and action on this! 👍🎉

**Comment by ameshkov** [comment]
Awesome, thank you for testing it!
